### ISSUE TITLE
Comparing uuids when joining tables with MS SQL database causes sql error when casting GUID to varchar

### ISSUE THREAD

**Issue by 37gears** [metabase/metabase#46148]
### Describe the bug

When using the drag and drop query builder joining tables with uuids as Keys, metabase generates sql that casts uuid to varchar with no length specified. This causes an error because MS SQL server's default varchar length is 30 and length must be 36 for uuids. See https://stackoverflow.com/questions/5628968/sql-server-insufficient-result-space-to-convert-uniqueidentifier-value-to-char.

This seems to have started following this fix - https://github.com/metabase/metabase/pull/45575) - worked as expected in version 50.11

### To Reproduce

1. Create a MS SQL database with one table as a uuid as primary key and another table with a foreign key of uuid to the first table 
2.  in metabase create a new question and join the two tables
3.  run the visualization
4. sql error occurs: An error occurred during the current command (Done status 0). Insufficient result space to convert uniqueidentifier value to char.

### Expected behavior

Expected SQL (v0.50.11)
SELECT
  TOP(1048575) "dbo"."TableA"."TableAId" AS "TableAId",
  ...
FROM
  "dbo"."TableA"
LEFT JOIN "dbo"."TableB" AS "TableB - TableAId" ON "dbo"."TableA"."TableAId" = "TableB - TableAId"."TableAId"

Generated SQL (v0.50.15)
SELECT
  TOP(1048575) "dbo"."TableA"."TableAId" AS "TableAId",
  ...
FROM
  "dbo"."TableA"
LEFT JOIN "dbo"."TableB" AS "TableB - TableAId" ON CAST("dbo"."TableA"."TableAId" AS varchar) = "TableB - TableAId"."TableAId"

### Information about your Metabase installation

```JSON
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36 Edg/126.0.0.0",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "21.0.4+7-LTS",
    "java.vendor": "Eclipse Adoptium",
    "java.vendor.url": "https://adoptium.net/",
    "java.version": "21.0.4",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "21.0.4+7-LTS",
    "os.name": "Windows Server 2022",
    "os.version": "10.0",
    "user.language": "en",
    "user.timezone": "America/New_York"
  },
  "metabase-info": {
    "databases": [
      "h2",
      "sqlserver"
    ],
    "hosting-env": "unknown",
    "application-database": "h2",
    "application-database-details": {
      "database": {
        "name": "H2",
        "version": "2.1.214 (2022-06-13)"
      },
      "jdbc-driver": {
        "name": "H2 JDBC Driver",
        "version": "2.1.214 (2022-06-13)"
      }
    },
    "run-mode": "prod",
    "plan-alias": "",
    "version": {
      "date": "2024-07-23",
      "tag": "v0.50.15",
      "hash": "c6697cf"
    },
    "settings": {
      "report-timezone": null
    }
  }
}
```

### Severity

blocking some users

**Comment by snoe** [comment]
I believe there's two issues here.

The one @37gears is showing should be fixed by https://github.com/metabase/metabase/issues/46558 in `50.20` as we should not be doing a cast at all in this situation. (I need to verify this still.)

The second issue, as they point out, is that sql server will need special handling to cast guids in the face of string operations (`contains`/`starts-with`/`ends-with`) which are relatively more rare.

**Comment by iethree** [comment]
🚀 This should also be released by [v0.51](https://github.com/metabase/metabase/milestone/231)
